src/sema/unused_variable.rs

```diff
@@ -0,0 +1,314 @@
+/// usage of the latter as well
+pub fn check_function_call(ns: &mut Namespace, exp: &Expression, symtable: &mut Symtable) {
+    match &exp {
+        Expression::InternalFunctionCall {loc: _, returns: _, function, args} |
+        Expression::ExternalFunctionCall {loc: _, returns: _, function, args, ..}  => {
+            for arg in args {
+                used_variable(ns, arg, symtable);
+            }
+            check_function_call(ns, function, symtable);
+        }
+
+        Expression::Constructor {loc: _, contract_no: _, constructor_no: _, args, ..} => {
+            for arg in args {
+                used_variable(ns, arg, symtable);
+            }
+        }
+
+        Expression::ExternalFunctionCallRaw {loc: _, ty:_, address: function, args, ..} => {
+            used_variable(ns, args, symtable);
+            check_function_call(ns, function, symtable);
+        }
+
+        Expression::ExternalFunction {loc: _, ty: _, address, function_no} => {
+            used_variable(ns, address, symtable);
+            if ns.functions[*function_no].is_accessor {
+                let body = ns.functions[*function_no].body[0].clone();
+                if let Statement::Return(_, exprs) = &body {
+                    used_variable(ns, &exprs[0], symtable);
+                }
+            }
+        }
+
+        Expression::Builtin(_, _, expr_type, args) => {
+            match expr_type {
+
+                Builtin::ArrayPush => {
+                    assigned_variable(ns, &args[0], symtable);
+                    if args.len() > 1 {
+                        used_variable(ns, &args[1], symtable);
+                    }
+                }
+
+                Builtin::ArrayPop => {
+                    used_variable(ns, &args[0], symtable);
+                }
+                _ => {}
+            }
+        }
+
+        Expression::DynamicArrayPush(_, array, _, arg) => {
+            assigned_variable(ns, array, symtable);
+            used_variable(ns, arg, symtable);
+        }
+
+        Expression::DynamicArrayPop(_, array, _) => {
+            used_variable(ns, array, symtable);
+        }
+
+        _ => {}
+    }
+}
+
+/// Marks as used variables that appear in an expression with right and left hand side.
+pub fn check_var_usage_expression(ns: &mut Namespace, left: &Expression, right: &Expression,
+                                  symtable: &mut Symtable) {
+    used_variable(ns, left, symtable);
+    used_variable(ns, right, symtable);
+}
+
+/// Generate warnings for unused varibles
+fn generate_unused_warning(loc: Loc, text: &String, notes: Vec<Note>) -> Box<Diagnostic> {
+    Box::new(Diagnostic {
+        level: Level::Warning,
+        ty: ErrorType::Warning,
+        pos: Some(loc),
+        message: text.clone(),
+        notes
+    })
+}
+
+/// Emit different warning types according to the state variable usage
+pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Box<Diagnostic>> {
```

Comment:
Why is the diagnositic in a box?

Reply:
I create the Diagnostic struct in function `generate_unused_warning` and it is returned to `emit_warning_local_variable`. There I return the value again to the original caller. I wanted to avoid memory copy between these three returns. I thought that using a smart pointer would be good.

Reply:
The only caller of `emit_warning_local_variable()` derefs the box with `*`, which effectively means unboxing it. Also using box for efficiency isn't the right thing to do: rust does a lot of inlining and tries to solve these issues for you.

Reply:
Fixed